In [1]:
import pandas as pd

# Customers Dataset

In [10]:
customers = pd.read_csv('data/olist_customers_dataset.csv')
customers.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

In [5]:
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB


In [26]:
customers.isna().sum()

customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

Rows = 99,441;
 Columns = 5;
 Key columns = customer_id, customer_unique_id;
 Missing values = None;
 Date columns = None;
 Unique IDs = customer_id (per order), customer_unique_id (per customer);
 Primary key = customer_id;
 Foreign key = customer_id → olist_orders_dataset

This dataset contains customer identifiers and geographical information
and is used to link customers with their orders.

# Geolocation Dataset

In [11]:
geolocation = pd.read_csv('data/olist_geolocation_dataset.csv')
geolocation.head()

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                         1037       -23.545621       -46.639292   
1                         1046       -23.546081       -46.644820   
2                         1046       -23.546129       -46.642951   
3                         1041       -23.544392       -46.639499   
4                         1035       -23.541578       -46.641607   

  geolocation_city geolocation_state  
0        sao paulo                SP  
1        sao paulo                SP  
2        sao paulo                SP  
3        sao paulo                SP  
4        sao paulo                SP

In [12]:
geolocation.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 5 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   geolocation_zip_code_prefix  1000163 non-null  int64  
 1   geolocation_lat              1000163 non-null  float64
 2   geolocation_lng              1000163 non-null  float64
 3   geolocation_city             1000163 non-null  object 
 4   geolocation_state            1000163 non-null  object 
dtypes: float64(2), int64(1), object(2)
memory usage: 38.2+ MB


In [41]:
geolocation.isna().sum()

geolocation_zip_code_prefix    0
geolocation_lat                0
geolocation_lng                0
geolocation_city               0
geolocation_state              0
dtype: int64

Rows = 1,000,163;
 Columns = 5;
 Key columns = geolocation_zip_code_prefix;
 Missing values = None;
 Date columns = None;
 Unique IDs = geolocation_zip_code_prefix (not strictly unique, repeated for multiple locations);
 Primary key = None;
 Foreign key = geolocation_zip_code_prefix → customers / sellers datasets

This dataset contains latitude and longitude information for ZIP code prefixes
and is used for mapping customer and seller locations.

# Order Items Dataset

In [14]:
order_items = pd.read_csv('data/olist_order_items_dataset.csv')
order_items.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

In [15]:
order_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB


In [28]:
order_items.isna().sum()

order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

Rows = 112,650;
 Columns = 7;
 Key columns = order_id, product_id, seller_id;
 Missing values = None;
 Date columns = shipping_limit_date;
 Unique IDs = order_id + order_item_id (together uniquely identify a row);
 Primary key = Composite key(order_id, order_item_id);
 Foreign key = order_id → orders dataset, product_id → products dataset, seller_id → sellers dataset

This dataset contains individual items within each order, including product, seller,
price, and shipping information.

# Order Payments Dataset

In [16]:
order_payments = pd.read_csv('data/olist_order_payments_dataset.csv')
order_payments.head()

order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45

In [17]:
order_payments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  object 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  object 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 4.0+ MB


In [29]:
order_payments.isna().sum()

order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64

Rows = 103,886;
 Columns = 5;
 Key columns = order_id;
 Missing values = None;
 Date columns = None;
 Unique IDs = order_id + payment_sequential (uniquely identify each payment);
 Primary key = Composite key(order_id, payment_sequential);
 Foreign key = order_id → orders dataset

This dataset contains payment details for orders, including payment type,
installments, and payment value.

# Order Reviews Dataset

In [18]:
order_reviews = pd.read_csv('data/olist_order_reviews_dataset.csv')
order_reviews.head()

review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   

   review_score review_comment_title  \
0             4                  NaN   
1             5                  NaN   
2             5                  NaN   
3             5                  NaN   
4             5                  NaN   

                              review_comment_message review_creation_date  \
0                                                NaN  2018-01-18 00:00:00   
1                                                NaN  2018-03-10 00:00:00   
2                                                NaN  2018-02-17 00:00:00   
3              Recebi bem antes do prazo estipulado.  2017-04-21 00:00:00   
4  Parabéns lojas lannister adorei comprar pela I...  2018-03-01 00:00:00   

  review_answer_timestamp  
0     2018-01-18 21:46:59  
1     2018-03-11 03:05:13  
2     2018-02-18 14:36:24  
3     2017-04-21 22:02:06  
4     2018-03-02 10:26:53

In [19]:
order_reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   review_id                99224 non-null  object
 1   order_id                 99224 non-null  object
 2   review_score             99224 non-null  int64 
 3   review_comment_title     11568 non-null  object
 4   review_comment_message   40977 non-null  object
 5   review_creation_date     99224 non-null  object
 6   review_answer_timestamp  99224 non-null  object
dtypes: int64(1), object(6)
memory usage: 5.3+ MB


In [30]:
order_reviews.isna().sum()

review_id                      0
order_id                       0
review_score                   0
review_comment_title       87656
review_comment_message     58247
review_creation_date           0
review_answer_timestamp        0
dtype: int64

Rows = 99,224;
 Columns = 7;
 Key columns = review_id, order_id;
 Missing values = review_comment_title, review_comment_message;
 Date columns = review_creation_date, review_answer_timestamp;
 Unique IDs = review_id;
 Primary key = review_id;
 Foreign key = order_id → orders dataset

This dataset contains customer reviews and ratings for orders,
including review scores and feedback timestamps.

# Orders Dataset

In [20]:
orders = pd.read_csv('data/olist_orders_dataset.csv')
orders.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00

In [21]:
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 non-null  object
dtypes: object(8)
memory usage: 6.1+ MB


In [31]:
orders.isna().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

Rows = 99,441;
 Columns = 8;
 Key columns = order_id, customer_id;
 Missing values = order_approved_at, order_delivered_carrier_date, order_delivered_customer_date;
 Date columns = order_purchase_timestamp, order_approved_at, order_delivered_carrier_date, order_delivered_customer_date, order_estimated_delivery_date;
 Unique IDs = order_id;
 Primary key = order_id;
 Foreign key = customer_id → customers dataset

This dataset contains order lifecycle information such as order status
and various order-related timestamps.

# Products dataset

In [22]:
products = pd.read_csv('data/olist_products_dataset.csv')
products.head()

product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0  
2             154.0               18.0                9.0              15.0  
3             371.0               26.0                4.0              26.0  
4             625.0               20.0               17.0              13.0

In [23]:
products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32341 non-null  object 
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.3+ MB


In [32]:
products.isna().sum()

product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

Rows = 32,951;
 Columns = 9;
 Key columns = product_id;
 Missing values = product_category_name, product_name_lenght, product_description_lenght, product_photos_qty, product_weight_g, product_length_cm, product_height_cm, product_width_cm;
 Date columns = None;
 Unique IDs = product_id;
 Primary key = product_id;
 Foreign key = product_id → order_items dataset

This dataset contains product-related information such as category,
physical dimensions, and product attributes.

# Sellers Dataset

In [24]:
sellers = pd.read_csv('data/olist_sellers_dataset.csv')
sellers.head()

seller_id  seller_zip_code_prefix  \
0  3442f8959a84dea7ee197c632cb2df15                   13023   
1  d1b65fc7debc3361ea86b5f14c68d2e2                   13844   
2  ce3ad9de960102d0677a81f5d0bb7b2d                   20031   
3  c0f3eea2e14555b6faeea3dd58c1b1c3                    4195   
4  51a04a8a6bdcb23deccc82b0b80742cf                   12914   

         seller_city seller_state  
0           campinas           SP  
1         mogi guacu           SP  
2     rio de janeiro           RJ  
3          sao paulo           SP  
4  braganca paulista           SP

In [25]:
sellers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3095 entries, 0 to 3094
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   seller_id               3095 non-null   object
 1   seller_zip_code_prefix  3095 non-null   int64 
 2   seller_city             3095 non-null   object
 3   seller_state            3095 non-null   object
dtypes: int64(1), object(3)
memory usage: 96.8+ KB


In [33]:
sellers.isna().sum()

seller_id                 0
seller_zip_code_prefix    0
seller_city               0
seller_state              0
dtype: int64

Rows = 3,095;
 Columns = 4;
 Key columns = seller_id;
 Missing values = None;
 Date columns = None;
 Unique IDs = seller_id;
 Primary key = seller_id;
 Foreign key = seller_id → order_items dataset

This dataset contains seller identifiers and geographical location details
used to link sellers with the products they sell.

# Product Category Name Dataset

In [36]:
product_category_name_translation = pd.read_csv('data/product_category_name_translation.csv')
product_category_name_translation.head()

product_category_name product_category_name_english
0            beleza_saude                 health_beauty
1  informatica_acessorios         computers_accessories
2              automotivo                          auto
3         cama_mesa_banho                bed_bath_table
4        moveis_decoracao               furniture_decor

In [37]:
product_category_name_translation.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71 entries, 0 to 70
Data columns (total 2 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   product_category_name          71 non-null     object
 1   product_category_name_english  71 non-null     object
dtypes: object(2)
memory usage: 1.2+ KB


In [39]:
product_category_name_translation.isna().sum()

product_category_name            0
product_category_name_english    0
dtype: int64

Rows = 71;
 Columns = 2;
 Key columns = product_category_name;
 Missing values = None;
 Date columns = None;
 Unique IDs = product_category_name;
 Primary key = product_category_name;
 Foreign key = product_category_name → products dataset

This dataset contains translations of product category names
from Portuguese to English.

olist_orders_dataset, olist_order_items_dataset, olist_customers_dataset, olist_product_dataset, and product_category_name_translation_dataset will be used for further analysis

# Data Cleaning 

Converting date columns to datetime in orders

In [59]:
date_cols = [
    'order_purchase_timestamp',
    'order_approved_at',
    'order_delivered_carrier_date',
    'order_delivered_customer_date',
    'order_estimated_delivery_date'
]

for col in date_cols:
    orders[col] = pd.to_datetime(orders[col], errors='coerce')

In [60]:
orders['order_status'].value_counts()

order_status
delivered     96478
shipped        1107
invoiced        314
processing      301
created           5
approved          2
Name: count, dtype: int64

The orders dataset does not contain any canceled or unavailable orders.

In [56]:
orders.dtypes

order_id                                 object
customer_id                              object
order_status                             object
order_purchase_timestamp         datetime64[ns]
order_approved_at                datetime64[ns]
order_delivered_carrier_date     datetime64[ns]
order_delivered_customer_date    datetime64[ns]
order_estimated_delivery_date    datetime64[ns]
dtype: object

order_items already has price (product price) and freight_value (shipping cost). Total item values needs to be calculated. A "total_item_price" column is added.

In [69]:
order_items['total_item_price'] = (
    order_items['price'] + order_items['freight_value']
)

In [64]:
order_items

order_id  order_item_id  \
0       00010242fe8c5a6d1ba2dd792cb16214              1   
1       00018f77f2f0320c557190d7a144bdd3              1   
2       000229ec398224ef6ca0657da4fc703e              1   
3       00024acbcdf0a6daa1e931b038114c75              1   
4       00042b26cf59d7ce69dfabb4e55b4fd9              1   
...                                  ...            ...   
112645  fffc94f6ce00a00581880bf54a75a037              1   
112646  fffcd46ef2263f404302a634eb57f7eb              1   
112647  fffce4705a9662cd70adb13d4a31832d              1   
112648  fffe18544ffabc95dfada21779c9644f              1   
112649  fffe41c64501cc87c801fd61db3f6244              1   

                              product_id                         seller_id  \
0       4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1       e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2       c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3       7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4       ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   
...                                  ...                               ...   
112645  4aa6014eceb682077f9dc4bffebc05b0  b8bc237ba3788b23da09c0f1f3a3288c   
112646  32e07fd915822b0765e448c4dd74c828  f3c38ab652836d21de61fb8314b69182   
112647  72a30483855e2eafc67aee5dc2560482  c3cfdc648177fdbbbb35635a37472c53   
112648  9c422a519119dcad7575db5af1ba540e  2b3e4a2a3ea8e01938cabda2a3e5cc79   
112649  350688d9dc1e75ff97be326363655e01  f7ccf836d21b2fb1de37564105216cc1   

        shipping_limit_date   price  freight_value  total_item_price  
0       2017-09-19 09:45:35   58.90          13.29             72.19  
1       2017-05-03 11:05:13  239.90          19.93            259.83  
2       2018-01-18 14:48:30  199.00          17.87            216.87  
3       2018-08-15 10:10:18   12.99          12.79             25.78  
4       2017-02-13 13:57:51  199.90          18.14            218.04  
...                     ...     ...            ...               ...  
112645  2018-05-02 04:11:01  299.99          43.41            343.40  
112646  2018-07-20 04:31:48  350.00          36.53            386.53  
112647  2017-10-30 17:14:25   99.90          16.95            116.85  
112648  2017-08-21 00:04:32   55.99           8.72             64.71  
112649  2018-06-12 17:10:13   43.00          12.79             55.79  

[112650 rows x 8 columns]

total_item_price is created (product price + freight)

In [70]:
products['product_category_name']

0                               perfumaria
1                                    artes
2                            esporte_lazer
3                                    bebes
4                    utilidades_domesticas
                       ...                
32946                     moveis_decoracao
32947    construcao_ferramentas_iluminacao
32948                      cama_mesa_banho
32949               informatica_acessorios
32950                      cama_mesa_banho
Name: product_category_name, Length: 32951, dtype: object

In [71]:
products['product_category_name'].isna().sum()

np.int64(610)

The product_category_name column contains 610 missing values which needs to be handled

In [73]:
products['product_category_name'] = (
    products['product_category_name'].fillna('unknown')
)

In [74]:
products['product_category_name'].isna().sum()

np.int64(0)

The missing values replaced with 'unknown'

Product categories needs to be translated to English by merging with the
product_category_name_translation dataset.

In [76]:
products = products.merge(
    product_category_name_translation,
    on='product_category_name',
    how='left'
)

In [77]:
products.head()

product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  \
0             225.0               16.0               10.0              14.0   
1            1000.0               30.0               18.0              20.0   
2             154.0               18.0                9.0              15.0   
3             371.0               26.0                4.0              26.0   
4             625.0               20.0               17.0              13.0   

  product_category_name_english  
0                     perfumery  
1                           art  
2                sports_leisure  
3                          baby  
4                    housewares

In [79]:
# Handling missing translations 
products['product_category_name_english'] = (
    products['product_category_name_english']
    .fillna('unknown')
)

In [80]:
products['product_category_name_english'].isna().sum()

np.int64(0)

product merged with product_category_name and missing values replaced with "unknown"

In [81]:
customers[['customer_state', 'customer_city']].isna().sum()

customer_state    0
customer_city     0
dtype: int64

The customers dataset already contains customer_state and customer_city.
No missing values were found, so no cleaning required

# Merging the datasets

In [82]:
df = order_items.merge(
    orders[['order_id', 'customer_id', 'order_purchase_timestamp']],
    on='order_id',
    how='inner'
)

In [83]:
df = df.merge(
    customers[['customer_id', 'customer_state']],
    on='customer_id',
    how='left'
)

In [85]:
df = df.merge(
    products[['product_id', 'product_category_name_english']],
    on='product_id',
    how='left'
)

In [93]:
df.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  total_item_price  \
0  2017-09-19 09:45:35   58.90          13.29             72.19   
1  2017-05-03 11:05:13  239.90          19.93            259.83   
2  2018-01-18 14:48:30  199.00          17.87            216.87   
3  2018-08-15 10:10:18   12.99          12.79             25.78   
4  2017-02-13 13:57:51  199.90          18.14            218.04   

                        customer_id order_purchase_timestamp customer_state  \
0  3ce436f183e68e07877b285a838db11a      2017-09-13 08:59:02             RJ   
1  f6dd3ec061db4e3987629fe6b26e5cce      2017-04-26 10:53:06             SP   
2  6489ae5e4333f3693df5ad4372dab6d3      2018-01-14 14:33:31             MG   
3  d4eb9395c8c0431ee92fce09860c5a06      2018-08-08 10:00:35             SP   
4  58dbd0b2d70206bf40e62cd34e84d795      2017-02-04 13:57:51             SP   

  product_category_name_english  
0                    cool_stuff  
1                      pet_shop  
2               furniture_decor  
3                     perfumery  
4                  garden_tools

Order items were merged with orders, customers, and products datasets.

Each row in the final dataset represents a single product item sold,
its order information, customer state, product category,
item price with freight value, and purchase timestamp.

In [115]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112101 entries, 0 to 112100
Data columns (total 14 columns):
 #   Column                         Non-Null Count   Dtype         
---  ------                         --------------   -----         
 0   order_id                       112101 non-null  object        
 1   order_item_id                  112101 non-null  int64         
 2   product_id                     112101 non-null  object        
 3   seller_id                      112101 non-null  object        
 4   shipping_limit_date            112101 non-null  object        
 5   price                          112101 non-null  float64       
 6   freight_value                  112101 non-null  float64       
 7   total_item_price               112101 non-null  float64       
 8   customer_id                    112101 non-null  object        
 9   order_purchase_timestamp       112101 non-null  datetime64[ns]
 10  customer_state                 112101 non-null  object        
 11  

In [92]:
df.shape

(112101, 12)

In [91]:
df.isna().sum()

order_id                         0
order_item_id                    0
product_id                       0
seller_id                        0
shipping_limit_date              0
price                            0
freight_value                    0
total_item_price                 0
customer_id                      0
order_purchase_timestamp         0
customer_state                   0
product_category_name_english    0
dtype: int64

In [141]:
df['order_purchase_timestamp'] = pd.to_datetime(
    df['order_purchase_timestamp'],
    errors='coerce'
)

In [143]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112101 entries, 0 to 112100
Data columns (total 12 columns):
 #   Column                         Non-Null Count   Dtype         
---  ------                         --------------   -----         
 0   order_id                       112101 non-null  object        
 1   order_item_id                  112101 non-null  int64         
 2   product_id                     112101 non-null  object        
 3   seller_id                      112101 non-null  object        
 4   shipping_limit_date            112101 non-null  object        
 5   price                          112101 non-null  float64       
 6   freight_value                  112101 non-null  float64       
 7   total_item_price               112101 non-null  float64       
 8   customer_id                    112101 non-null  object        
 9   order_purchase_timestamp       112101 non-null  datetime64[ns]
 10  customer_state                 112101 non-null  object        
 11  

Additional time-based features - order_month and order_day
created for further analysis.

In [144]:
df.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  total_item_price  \
0  2017-09-19 09:45:35   58.90          13.29             72.19   
1  2017-05-03 11:05:13  239.90          19.93            259.83   
2  2018-01-18 14:48:30  199.00          17.87            216.87   
3  2018-08-15 10:10:18   12.99          12.79             25.78   
4  2017-02-13 13:57:51  199.90          18.14            218.04   

                        customer_id order_purchase_timestamp customer_state  \
0  3ce436f183e68e07877b285a838db11a      2017-09-13 08:59:02             RJ   
1  f6dd3ec061db4e3987629fe6b26e5cce      2017-04-26 10:53:06             SP   
2  6489ae5e4333f3693df5ad4372dab6d3      2018-01-14 14:33:31             MG   
3  d4eb9395c8c0431ee92fce09860c5a06      2018-08-08 10:00:35             SP   
4  58dbd0b2d70206bf40e62cd34e84d795      2017-02-04 13:57:51             SP   

  product_category_name_english  
0                    cool_stuff  
1                      pet_shop  
2               furniture_decor  
3                     perfumery  
4                  garden_tools

In [145]:
df[['order_purchase_timestamp']].dtypes

order_purchase_timestamp    datetime64[ns]
dtype: object

In [146]:
df['order_purchase_timestamp'].isna().sum()

np.int64(0)

In [147]:
df['order_id'].isin(
    orders[orders['order_status'].isin(['canceled','unavailable'])]['order_id']
).sum()

np.int64(0)

In [148]:
df['product_category_name_english'].isna().sum()

np.int64(0)

In [149]:
df['product_category_name_english'].unique()[:5]

array(['cool_stuff', 'pet_shop', 'furniture_decor', 'perfumery',
       'garden_tools'], dtype=object)

In [150]:
df.shape

(112101, 12)

In [151]:
df.columns

Index(['order_id', 'order_item_id', 'product_id', 'seller_id',
       'shipping_limit_date', 'price', 'freight_value', 'total_item_price',
       'customer_id', 'order_purchase_timestamp', 'customer_state',
       'product_category_name_english'],
      dtype='object')

All preprocessing checkpoints verified.

In [152]:
df.to_csv(
    'data/final_dataset.csv',
    index=False
)

In [153]:
saved_df = pd.read_csv('data/final_dataset.csv')
saved_df.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  total_item_price  \
0  2017-09-19 09:45:35   58.90          13.29             72.19   
1  2017-05-03 11:05:13  239.90          19.93            259.83   
2  2018-01-18 14:48:30  199.00          17.87            216.87   
3  2018-08-15 10:10:18   12.99          12.79             25.78   
4  2017-02-13 13:57:51  199.90          18.14            218.04   

                        customer_id order_purchase_timestamp customer_state  \
0  3ce436f183e68e07877b285a838db11a      2017-09-13 08:59:02             RJ   
1  f6dd3ec061db4e3987629fe6b26e5cce      2017-04-26 10:53:06             SP   
2  6489ae5e4333f3693df5ad4372dab6d3      2018-01-14 14:33:31             MG   
3  d4eb9395c8c0431ee92fce09860c5a06      2018-08-08 10:00:35             SP   
4  58dbd0b2d70206bf40e62cd34e84d795      2017-02-04 13:57:51             SP   

  product_category_name_english  
0                    cool_stuff  
1                      pet_shop  
2               furniture_decor  
3                     perfumery  
4                  garden_tools

The cleaned and merged master dataset exported as a CSV file